modular-framework/module.h

```diff
@@ -0,0 +1,205 @@
+
+#define subsystem_constructor(name) 				\
+	do {							\
+		rwlock_init(&subsystem(name).lock);		\
+		INIT_LIST_HEAD(&subsystem(name).modules);	\
+		subsystem(name).active = NULL;			\
+	} while(0)
+
+#define SUBSYSTEM_CONSTRUCTOR(name) 				\
+	static void __attribute__((constructor(101)))		\
+		name ## _subsystem_constructor(void)
+
+#define subsystem_lock(access, name)				\
+	rwlock_ ##access## _lock(&subsystem(name).lock)
+
+#define subsystem_unlock(access, name)				\
+	rwlock_ ##access## _unlock(&subsystem(name).lock)
+
+#define subsystem_foreach_module(name, mod)			\
+	list_for_each_entry(mod, &subsystem(name).modules, list)
+
+#define MODULE_CLASS(subsystem)					\
+	struct subsystem ## _module {				\
+		struct list_head list;				\
+		const char *name;				\
+		void *handler; /* DSO */			\
+		int (*init_local)(void);			\
+		int (*term_local)(void);			\
+		int (*init_global)(void);			\
+		int (*term_global)(void);			\
+
+/* Base class to all inherited subsystem module classes */
+typedef MODULE_CLASS(base) } module_base_t;
+
+/* Module constructors should be late than subsystem constructors,
+ * in statically linked scenarios (both subsystems and modules are
+ * linked statically). thus the priority 102 compared to the above
+ * subsystem constructor priority 101.
+ */
+#define MODULE_CONSTRUCTOR(name) 				\
+	static void __attribute__((constructor(102)))		\
+		name ## _module_constructor(void)
+
+/* All subsystems' initialization and termination routines are
+ * the same, provide template to instantiation.
+ */
+#define SUBSYSTEM_INITERM_TEMPLATE(subs, method, print)		\
+static inline int subs ## _subsystem ##_## method(void)		\
+{								\
+	module_base_t *mod = NULL;				\
+								\
+	subsystem_lock(read, subs);				\
+	subsystem_foreach_module(subs, mod) {			\
+		int result = mod->method ? mod->method() : -1;	\
+		if (result < 0) {				\
+			subsystem_unlock(read, subs);		\
+			print("error %d to %s subsystem %s "	\
+			      "module %s.\n", result, # method, \
+			      subsystem(subs).name, mod->name);	\
+			return result;				\
+		}						\
+	}							\
+	subsystem_unlock(read, subs);				\
+	return 0;						\
+}
+
+/* Subsystem Modules Registration
+ *
+ * subsystem_register_module() are called by all modules in their
+ * constructors, whereas the modules could be:
+ *
+ * 1) Linked statically or dynamically, and are loaded by program
+ *    loader (execve) or dynamic linker/loader (ld.so)
+ *
+ *    subsystem_register_module() should complete the whole
+ *    registration session and link the module into subsystem's
+ *    module array.
+ *
+ * 2) Loaded by a module loader in runtime with libdl APIs
+ *
+ *    The whole registration session needs to be split to aim the
+ *    module loader to properly handle dlopen() returns, and save
+ *    the DSO handler into module's data structure.
+ *
+ *    The module loader should program in this way:
+ *	module_loader_start();
+ *	......
+ * 	for each module
+ *		handler = dlopen(module)
```

Comment:
This patch series did not implement a plugin loader (dlopen()), an example can be found:
https://github.com/heyi-linaro/pure-interface/blob/master/module-loader.c